src/osd/osd_types.h

```diff
@@ -4366,47 +4451,60 @@ class pg_missing_set : public pg_missing_const_i {
     if (e.prior_version == eversion_t() || e.is_clone()) {
       // new object.
       if (is_missing_divergent_item) {  // use iterator
-	rmissing.erase((missing_it->second).need.version);
-	missing_it->second = item(e.version, eversion_t(), e.is_delete());  // .have = nil
-      } else  // create new element in missing map
-	missing[e.soid] = item(e.version, eversion_t(), e.is_delete());     // .have = nil
+	      rmissing.erase((missing_it->second).need.version);
+        // .have = nil
+	missing_it->second = item(e.version, eversion_t(), e.is_delete());
+	missing_it->second.clean_regions.mark_fully_dirty();
+      } else {
+         // create new element in missing map
+         // .have = nil
+        missing[e.soid] = item(e.version, eversion_t(), e.is_delete());
+        missing[e.soid].clean_regions = e.clean_regions;
+        missing[e.soid].clean_regions.mark_fully_dirty();
+      }
     } else if (is_missing_divergent_item) {
       // already missing (prior).
       rmissing.erase((missing_it->second).need.version);
       (missing_it->second).need = e.version;  // leave .have unchanged.
-      missing_it->second.set_delete(e.is_delete());
+      (missing_it->second).set_delete(e.is_delete());
+      (missing_it->second).clean_regions.merge(e.clean_regions);
     } else {
       // not missing, we must have prior_version (if any)
       ceph_assert(!is_missing_divergent_item);
       missing[e.soid] = item(e.version, e.prior_version, e.is_delete());
+      missing[e.soid].clean_regions = e.clean_regions;
     }
     rmissing[e.version.version] = e.soid;
     tracker.changed(e.soid);
   }
 
   void revise_need(hobject_t oid, eversion_t need, bool is_delete) {
-    if (missing.count(oid)) {
-      rmissing.erase(missing[oid].need.version);
-      missing[oid].need = need;            // no not adjust .have
-      missing[oid].set_delete(is_delete);
+    auto p = missing.find(oid);
+    if (p != missing.end()) {
+      rmissing.erase((p->second).need.version);
+      (p->second).need = need;          // no not adjust .have
+      (p->second).set_delete(is_delete);
+      (p->second).clean_regions.mark_fully_dirty();
     } else {
       missing[oid] = item(need, eversion_t(), is_delete);
+      missing[oid].clean_regions.mark_fully_dirty();
     }
     rmissing[need.version] = oid;
 
     tracker.changed(oid);
   }
 
   void revise_have(hobject_t oid, eversion_t have) {
-    if (missing.count(oid)) {
+    auto p = missing.find(oid);
+    if (p != missing.end()) {
       tracker.changed(oid);
-      missing[oid].have = have;
+      (p->second).have = have;
     }
   }
 
   void add(const hobject_t& oid, eversion_t need, eversion_t have,
-	   bool is_delete) {
-    missing[oid] = item(need, have, is_delete);
+	   bool is_delete, bool make_dirty = true) {
+    missing[oid] = item(need, have, is_delete, make_dirty, have == eversion_t());
```

Comment:
using `have == eversion_t()` is a bit confusing here, maybe add a comment to explain it

Reply:
ok